xclim/indices/_agro.py

```diff
@@ -882,6 +883,188 @@ def water_budget(
+    """
+    # Unit conversion.
+    pram = rate2amount(pr, out_units="mm")
+    thresh_wet_start = convert_units_to(thresh_wet_start, pram)
+    thresh_dry_start = convert_units_to(thresh_dry_start, pram)
+    thresh_dry_end = convert_units_to(thresh_dry_end, pram)
+
+    last_doy = pram.indexes["time"][-1].strftime("%m-%d")
+
+    # Eliminate negative values.
+    pram = xarray.where(pram < 0, 0, pram)
+    pram.attrs["units"] = "mm"
+
+    # should we flag date_min_end  < date_max_start?
+    def _get_first_run(run_positions, start_date, end_date):
+        run_positions = select_time(run_positions, date_bounds=(start_date, end_date))
+        first_start = run_positions.argmax("time")
+        return xarray.where(
+            first_start != run_positions.argmin("time"), first_start, np.NaN
+        )
+
+    # Find the start of the rain season
+    def _get_first_run_start(pram, window_dry=window_dry_start):
+        pram = select_time(pram, date_bounds=(date_min_start, last_doy))
+
+        # First condition: Start with enough precipitation
+        da_start = pram.rolling({"time": window_wet_start}).sum() >= thresh_wet_start
+
+        # Second condition: No dry period after
+        if method_dry_start == "per_day":
+            da_stop = pram <= thresh_dry_start
+        elif method_dry_start == "total":
+            da_stop = pram.rolling({"time": window_dry}).sum() <= thresh_dry_start
+            # equivalent to rolling forward in time instead, i.e. end date will be at beginning of dry run
+            da_stop = da_stop.shift({"time": -(window_dry - 1)}, fill_value=False)
+            window_dry = 1
+
+        # First and second condition combined in a run length
+        run_positions = (
+            rl.rle_events(da_start, 1, da_stop, window_dry, "first")
+            >= window_not_dry_start + window_wet_start
+        )
+
+        return _get_first_run(run_positions, date_min_start, date_max_start)
+
+    # Find the end of the rain season
+    def _get_first_run_end(pram):
+        if method_dry_end == "per_day":
+            da_stop = pram <= thresh_dry_end
+            run_positions = rl.rle(da_stop, index="first") >= window_dry_end
+        elif method_dry_end == "total":
+            da_stop = pram.rolling({"time": window_dry_end}).sum() <= thresh_dry_end
+            run_positions = da_stop
+        return _get_first_run(run_positions, date_min_end, date_max_end)
+
+    # Get start, end and length of rain season. Written as a function so it can be resampled
+    def _get_rain_season(pram):
+        start = _get_first_run_start(pram)
+
+        # masking value before  start of the season (end of season should be after)
+        start_ind = xarray.where(start.notnull(), start.astype(int), -1)
+        mask = xarray.where(
+            pram.time == pram.isel(time=start_ind).time, 1, np.NaN
+        ).ffill("time")
+        mask[{"time": start_ind}] = np.NaN
+        mask = mask.notnull()
```

Comment:
```suggestion
    # Get valid integer indexer of the day after the first run starts
    start_ind = (start + 1).fillna(-1).astype(int)
    mask = (pram * np.NaN) 
    # Put "True" on the day of run start
    mask[{'time': start_ind}] = 1
    # Mask back points without runs, propagate the True
    mask = mask.where(start.notnull()).ffill('time')
    mask = mask.notnull()
```

I tried something to make this code clearer and shorter....

I assumed  `start + 1` is safe : the need for a stop condition makes it so that `start` can't be the last index. no ?

Reply:
I think it's fine yes. Maybe we should add:
```
start_ind = xr.where(start_ind < len(pram.time), start_ind,  -1)
```
to be on the safe side. 

Also, I think: 
```
mask = mask.where(start.notnull()).ffill('time')
```
could just be 
```
mask = mask.ffill('time')
```
at this point, it is also non-NaN only at places where start is non-NaN because we used : `start_ind = (start + 1).fillna(-1).astype(int)`